### ISSUE TITLE
Clang-13 incorrectly selects std::apply() where user's apply() in local namespace exists

### ISSUE THREAD

**Issue by ikitayama** [llvm/llvm-project#53084]
We found with `-std=c++17` option used Clang can not compile simple program shown below:

```
#include <tuple>
#include <vector>

namespace foo
{

void
bar()
{
}

std::vector< double >
apply( const std::vector< double >&, const std::vector< double >& )
{
  return {};
}

void
run()
{
  std::vector< double > vec;

  bar();                  // OK
  apply( vec, vec );      // Tries to use std::apply() from <tuple>, which causes an error with the arguments
  foo::apply( vec, vec ); // OK
}
} // end of namespace foo

int
main()
{
  foo::run();
}
```

**Comment by ikitayama** [comment]
https://github.com/nest/nest-simulator/issues/2254 for the details.

**Comment by ikitayama** [comment]
Cc: @hakonsbm, @heplesser

**Comment by JVApen** [comment]
Ain't this an unfortunate case of ADL where a better match is found because of the forwarding references in std::apply?
If you make `vec` const, you'll most likely get the right behavior.

**Comment by zero9178** [comment]
Yes this is almost certainly due to ADL. Clang, GCC and MSVC agree: https://godbolt.org/z/K6aejW9zq
Both arguments of `std::apply` are specified as universal references which therefore produce a function that after template substitution precisely matches the input parameters (`std::vector<double>&` in this case) and is therefore the better candidate.

A bit of a reduced test case: https://godbolt.org/z/x9sWj5hYc. Here you can see that `bar(b)` calls `A::bar<A::B&>` as after template substitution its a perfect match of input parameters. `foo::bar` would require an extra type conversion (`A::B&` to `const A::B&`). If we instead do `bar(std::as_const(b))` the input parameter is now `const A::B&` and `foo::bar` gets picked as it is a perfect match, and not a template. 

So to summarize, while unfortunate, this is the expected behaviour due to ADL and the fact that it happens in C++17 is due to C++17 having added `std::apply`. The correct fix is basically to fully qualify the call via `foo::apply` as you showed.

An easy and nice blog post about how C++ resolves a function call can be seen here: https://preshing.com/20210315/how-cpp-resolves-a-function-call/
